task/monorail-server/src/bsp/medusa_a.rs

```diff
@@ -0,0 +1,528 @@
+            seq,
+        };
+
+        out.reinit()?;
+        Ok(out)
+    }
+
+    pub fn reinit(&mut self) -> Result<(), VscError> {
+        ringbuf_entry!(Trace::Reinit);
+        self.vsc7448.init()?;
+
+        // By default, the SERDES6G are grouped into 4x chunks for XAUI,
+        // where a single DEV10G runs 4x SERDES6G at 2.5G.  This leads to very
+        // confusing behavior when only running a few SERDES6G: in particularly,
+        // we noticed that SERDES6G_14 seemed to depend on SERDES6G_12.
+        //
+        // We're never using this "lane sync" feature, so disable it everywhere.
+        for i in 0..=1 {
+            self.vsc7448.modify(
+                VAUI0().VAUI_CHANNEL_CFG().VAUI_CHANNEL_CFG(i),
+                |r| {
+                    r.set_lane_sync_ena(0);
+                },
+            )?;
+            self.vsc7448.modify(
+                VAUI1().VAUI_CHANNEL_CFG().VAUI_CHANNEL_CFG(i),
+                |r| {
+                    r.set_lane_sync_ena(0);
+                },
+            )?;
+        }
+
+        // We must disable frame copying before configuring ports; otherwise, a
+        // rare failure mode can result in queues getting stuck (forever!).  We
+        // disable frame copying by enabling VLANs, then removing all ports from
+        // them!
+        //
+        // (ports will be added back to VLANs after configuration is done, in
+        // the call to `configure_vlan_semistrict` below)
+        //
+        // The root cause is unknown, but we suspect a hardware race condition
+        // in the switch IC; see this issue for detailed discussion:
+        // https://github.com/oxidecomputer/hubris/issues/1399
+        self.vsc7448.configure_vlan_none()?;
+
+        // Reset internals
+        self.vsc8504 = Vsc8504::empty();
+        self.front_io_speed = [Speed::Speed1G; 2];
+
+        self.phy_vsc8504_init()?;
+
+        self.vsc7448.configure_ports_from_map(&PORT_MAP)?;
+        self.vsc7448.configure_vlan_semistrict()?;
+        self.vsc7448_postconfig()?;
+
+        // Some front IO boards have a faulty oscillator driving the PHY,
+        // causing its clock to misbehave some fraction of (re-)boots. Init
+        // the PHY in a loop, requesting the sequencer to reset as much as
+        // necessary to try and correct the problem.
+        let mut osc_good = false;
+
+        while self.vsc8562.is_some() && !osc_good {
+            self.phy_vsc8562_init()?;
+
+            osc_good = self.is_front_io_link_good()?;
+
+            // Notify the sequencer about the state of the oscillator. If the
+            // oscillator is good any future resets of the PHY do not require a
+            // full power cycle of the front IO board.
+            self.seq
+                .set_front_io_phy_osc_state(osc_good)
+                .map_err(|e| VscError::ProxyError(e.into()))?;
+
+            if !osc_good {
+                ringbuf_entry!(Trace::FrontIoPhyOscillatorBad)
```

Comment:
**question:** Should this only log once to avoid flooding the ringbuf in case resetting the sequencer isn't helping with the oscillator state? (And actually, should there be some limit where the software just gives up and decides that the oscillator is entirely unworkable?)

Reply:
This is not something I'm concerned about right now given our experience with the problematic part thus far. Sometime it will come up quickly, other times after several more power cycles. We've yet to encounter a case where it did not come up, and should that happen the impact is large enough that we'd need to take physical action with the switch anyway so I believe it is harmless to continue to cycle the board.

Reply:
@aapoalas  it's also worth noting that our ringbufs have a de-duplication feature where, if we record entries that are equal multiple times in a row, it just increments the event's counter rather than recording a new entry. so, if there aren't any other entries recorded in between these, it isn't really a big concern.